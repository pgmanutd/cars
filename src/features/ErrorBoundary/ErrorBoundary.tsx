import React, { ErrorInfo } from 'react';
import Typography from '@material-ui/core/Typography';
import _noop from 'lodash/fp/noop';

export type ErrorBoundaryProps = {};

export interface ErrorBoundaryState {
  hasError: boolean;
}

class ErrorBoundary extends React.PureComponent<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  constructor(props: ErrorBoundaryProps) {
    super(props);

    this.state = { hasError: false };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({ hasError: true });

    // TODO: Replace _noop with logToSever(error, errorInfo);
    _noop(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      return (
        <section data-testid="ErrorBoundary">
          <Typography variant="h4" gutterBottom>
            Something went wrong.
          </Typography>
          <Typography variant="body1">
            Refresh the page or try again later. If this error persists, Please
            contact the administrator.
          </Typography>
        </section>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
