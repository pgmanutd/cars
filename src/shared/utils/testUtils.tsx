import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { render } from '@testing-library/react';

export const renderWithRouter = (
  ui: React.ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: MemoryHistory } = {},
) => ({
  ...render(<Router history={history}>{ui}</Router>),
  history,
});
