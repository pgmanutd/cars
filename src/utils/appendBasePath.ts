const appendBasePath = (basePath: string, path: string) =>
  `${basePath}?${path}`;

export default appendBasePath;
