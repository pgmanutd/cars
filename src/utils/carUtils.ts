import { Car } from 'features/CarInfo/types';

export const getCarName = ({ manufacturerName, modelName }: Car) =>
  `${manufacturerName} ${modelName}`;

export const getCarFeatures = (
  { stockNumber, mileage, fuelType, color }: Car,
  { stockNumberPrefix = '' }: { stockNumberPrefix?: string } = {
    stockNumberPrefix: '',
  },
) =>
  `${stockNumberPrefix}${stockNumber} - ${mileage?.number} ${mileage?.unit} - ${fuelType} - ${color}`;
