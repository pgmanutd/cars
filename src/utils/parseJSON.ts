const parseJSON = <T>(stringifiedJSON: string | null, fallbackValue: T) => {
  try {
    if (!stringifiedJSON) {
      return fallbackValue;
    }

    return JSON.parse(stringifiedJSON);
  } catch (e) {
    return fallbackValue;
  }
};

export default parseJSON;
