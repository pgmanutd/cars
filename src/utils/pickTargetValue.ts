import React from 'react';

const pickTargetValue = (event: React.ChangeEvent<{ value: unknown }>) =>
  event.target.value;

export default pickTargetValue;
