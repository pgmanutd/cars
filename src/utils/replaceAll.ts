export type ReplaceAllMappedObject = { [key: string]: string };

const replaceAll = (
  value: string,
  mappedObject: ReplaceAllMappedObject = {},
) => {
  const keys = Object.keys(mappedObject);

  if (!keys.length) {
    return value;
  }

  const regex = new RegExp(keys.join('|'), 'gi');

  return value.replace(regex, (matched) => {
    return mappedObject[matched];
  });
};

export default replaceAll;
