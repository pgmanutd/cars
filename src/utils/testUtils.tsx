import React from 'react';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    routerConfig: {
      path = '/',
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    },
  }: {
    routerConfig: { path?: string; route?: string; history?: MemoryHistory };
  } = { routerConfig: {} },
) => ({
  ...render(
    <HelmetProvider>
      <Router history={history}>
        <Route
          path={path}
          render={routeProps => React.cloneElement(ui, { ...routeProps })}
        />
      </Router>
    </HelmetProvider>,
  ),
  history,
});
