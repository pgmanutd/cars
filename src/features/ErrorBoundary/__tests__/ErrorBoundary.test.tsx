import React from 'react';

import { renderWithTranslate } from 'utils/testUtils';

import ErrorBoundary from '../ErrorBoundary';

describe('<ErrorBoundary />', () => {
  beforeEach(() => {
    spyOn(console, 'error');
  });

  const setup = ({ shouldRenderChildComponentWithError = true } = {}) => {
    const ErrorThrower = () => {
      throw new Error('I crashed!');
    };

    const { renderResult } = renderWithTranslate(
      <ErrorBoundary>
        {shouldRenderChildComponentWithError && <ErrorThrower />}
        <div>Some child component</div>
      </ErrorBoundary>,
    );

    return {
      renderResult,
      ErrorThrower,
    };
  };

  describe('render the component and matches it against stored snapshot', () => {
    it('when some error is thrown from any child component', () => {
      const { renderResult } = setup();

      expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it('when no error is thrown from any child component', () => {
      const { renderResult } = setup({
        shouldRenderChildComponentWithError: false,
      });

      expect(renderResult.asFragment()).toMatchSnapshot();
    });
  });
});
