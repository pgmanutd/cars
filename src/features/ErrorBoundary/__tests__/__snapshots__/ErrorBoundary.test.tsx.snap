// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> render the component and matches it against stored snapshot when no error is thrown from any child component 1`] = `
<DocumentFragment>
  <div>
    Some child component
  </div>
</DocumentFragment>
`;

exports[`<ErrorBoundary /> render the component and matches it against stored snapshot when some error is thrown from any child component 1`] = `
<DocumentFragment>
  <section
    data-testid="ErrorBoundary"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom"
    >
      Something went wrong.
    </h4>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Refresh the page or try again later. If this error persists, Please contact the administrator.
    </p>
  </section>
</DocumentFragment>
`;
